pkg/mimirtool/client/backfill.go

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+func (c *MimirClient) Backfill(ctx context.Context, blocks []string) error {
+	// Upload each block
+	var succeeded, failed, alreadyExists int
+
+	for _, b := range blocks {
+		logctx := logrus.WithFields(logrus.Fields{"path": b})
+		if err := c.backfillBlock(ctx, b, logctx); err != nil {
+			if errors.Is(err, errConflict) {
+				logctx.Warning("block already exists on the server")
+				alreadyExists++
+			} else {
+				logctx.WithField("error", err).Error("failed uploading block")
+				failed++
+			}
+			continue
+		}
+
+		// no logging, backfillBlock already logged result.
+		succeeded++
+	}
+
+	logrus.WithFields(logrus.Fields{"succeeded": succeeded, "already_exists": alreadyExists, "failed": failed}).Info("finished uploading blocks")
+
+	if failed > 0 {
+		return fmt.Errorf("blocks failed to upload %d block(s)", failed)
+	}
+
+	return nil
+}
+
+// drainAndCloseBody drains and closes the body to let the transport reuse the connection.
+func drainAndCloseBody(resp *http.Response) {
+	_, _ = io.Copy(io.Discard, resp.Body)
+	_ = resp.Body.Close()
+}
+
+func (c *MimirClient) backfillBlock(ctx context.Context, blockDir string, logctx *logrus.Entry) error {
+	// blockMeta returned by getBlockMeta will have thanos.files section pre-populated.
+	blockMeta, err := getBlockMeta(blockDir)
+	if err != nil {
+		return err
+	}
+
+	blockID := blockMeta.ULID.String()
+	logctx = logctx.WithFields(logrus.Fields{"block": blockID})
+
+	logctx.WithField("file", "meta.json").Info("making request to start block upload")
+
+	blockUploadEndpointPrefix := path.Join("/api/v1/upload/block", url.PathEscape(blockID))
+
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(blockMeta); err != nil {
+		return errors.Wrap(err, "failed to JSON encode payload")
+	}
+	resp, err := c.doRequest(blockUploadEndpointPrefix, http.MethodPost, buf, int64(buf.Len()))
+	if err != nil {
+		return errors.Wrap(err, "request to start block upload failed")
+	}
+	drainAndCloseBody(resp)
+	if resp.StatusCode/100 != 2 {
+		return fmt.Errorf("request to start block upload failed, with HTTP status %d %s",
```

Comment:
We're shallowing the actual error message. My main concern is that showing just the status code to the user may not be much useful.

_Same comment applies on other places._

Reply:
`client.doRequest` already logs response from the server.

Reply:
Actually `doRequest` already checks the response, we don't need to check it again here. 🤔

Reply:
I've removed extra status code checks.